### ISSUE TITLE
Page numbers are displayed as "1" after about page 1,000

### ISSUE THREAD

**Issue by archfrog** [typst/typst#677]
Hi,

I love Typst but decided to push it a bit before I lean too strongly on it.  To this purpose, I created a huge test case that is a little more than a thousand A4 pages in size.  All in all, Typst seems to handle a huge book well, but there appears to be an issue with page numbers past the 63rd heading (after page 966) in the book: They are reset to 1 and remains so to the end of the book.

I've attached a ZIP archive containing a Python3 script to generate the book and the templates used.  To build, simply issue the command `typst compile huge.typ`.  I was using the most recent GitHub version (which I built) on Windows x64.  On Linux (using the same version) the problem does not appear, so it seems to be something specific to Windows x64.  I've attached the PDF too.

[typst-huge-page-number-issue.zip](https://github.com/typst/typst/files/11183801/typst-huge-page-number-issue.zip)
[huge.pdf](https://github.com/typst/typst/files/11183840/huge.pdf)

Cheers,
Archfrog

**Comment by archfrog** [comment]
Thank you for looking into the issue!  You `huge.pdf` file looks just like the version I built on Linux (where the `Crimson Pro` font isn't installed).  If I disable the font selection code (`#set text(font: "Crimson Pro")`), I get the same result as you on Windows too.

I just redid the test on Windows x64 using Typst v0.1 with the `Crimson Pro` font active and the issue is there as well.  So I *guess* that it is present on both Windows and Linux, but only if the `Crimson Pro` font is installed on the system.

I am thinking perhaps the problem, when using the `Crimson Pro` font, is that some bounding box gets too small so only the initial `1` of `1000`+ gets displayed?

By the way, `Crimson Pro` is a free font that can be fetched from Google Fonts and elsewhere.

**Comment by archfrog** [comment]
Your idea sounds very plausible.

Also, my idea about a too small bounding box does not make any sense when I think about it: The problem occurs already on page 967 (a page number without any ones in it), so I think your idea is much more likely.

This particular case is definitely not a show stopper for me so I'm happy to wait for something to happen down the road, perhaps an update suddenly fixes it, and then I can retry.  If I could mark this issue as a low-priority issue, I would.

I am very satisfied with Typst and "huge books", just something I wanted to make sure it could handle in a case future book ever grows to that size.

**Comment by archfrog** [comment]
FYI:
I just converted a partial book (48 A5 pages) from Word to Typst, using `pandoc --columns=132 --standalone --output=foo.typ foo.docx`, edited it to use the `Crimson Pro` font, and immediately ran into this issue.  This was also with the `Crimson Pro` free font.

I then tried building the same partial book using Adobe's `Minion Pro` font (which I've purchased) and there were no issues at all (as far as I can see), in particular the last page numbers are correct in this version.

So my guess is that the `Crimson Pro` font is somehow invalid or doesn't do things the proper way.  I have a few times noticed squares, indicating missing characters, when I used the `Crimson Pro` font, so I guess I'll have to continue my search for a complete, good, reliable free font that looks pleasing to the eye.

**Comment by archfrog** [comment]
I tried these combinations:

1. Typst v0.1 with `CrimsonPro.zip`: The issue persists.
2. Typst v0.1 with `CrimsonPro_no_font_features.zip`: The issue persists.
3. Typst v0.2 with `CrimsonPro.zip`: The issue persists.
4. Typst v0.2 with `CrimsonPro_no_font_features.zip`: The issue persists.
5. Typst v0.1 with `Minion Pro` (a professional Adobe font): No issue.
6. Typst v0.2 with `Minion Pro` (-"-): No issue.

I actually tried all of the above with both the partial book of 48 A5 pages and the "huge.typ" sample, with bold, italics, and underline disabled, so it was some 12 cases, all of them pointing towards an issue with the `Crimson Pro` font.

One thing that I am uncertain of is if Typst supports "dynamic fonts".  The Crimson Pro archive from Google Fonts includes both "dynamic" and "static" fonts.  There are two files of dynamic fonts and 16 files of static fonts, where the two dynamic files supposedly include all the font information present in the 16 static files.

Just as a precaution, I tried uninstalling all Crimson Pro fonts in my Windows system and then I installed both of the "dynamic" font files.  And, voila, suddenly the page numbers are correct, all the way to the last page.

I know next to nothing about fonts, but perhaps a note should be added to the install instructions that Typst prefers dynamic fonts over static fonts.  Or, perhaps there is some sort of issue with the static fonts for Crimson Pro.  I am going to use the dynamic fonts until, if ever, I run into problems with them.

**To summarize:**
If using "static" fonts, Typst generates wrong page numbers on the last few pages (they always get the page number 1).
If using "dynamic" fonts, there are no issues with Typst v0.2 (I didn't try it out with v0.1 as I just updated my local copy).

Here are the exact fonts that I use (I should have attached those long ago):

* [Crimson_Pro.zip](https://github.com/typst/typst/files/11238860/Crimson_Pro.zip)

**Comment by archfrog** [comment]
FYI:
I just tried out your tiny test that recreates the issue and there's no issue with the currently installed dynamic Crimson Pro fonts.

**Comment by archfrog** [comment]
Sigh.  I was working on standardizing the layout of my partial book and suddenly I discovered the issue again.  This time with the dynamic Crimson Pro font.  If I go back to Minion Pro, it works as expected.  So the good news weren't that good after all, only a somewhat hard-to-predict issue.

I have made a copy of my working files, but I cannot upload them here as we're talking about a book I'm working on.

I think your tiny test case is the best test case there is, so there shouldn't be any need for a copy of my partial book.

**Comment by laurmaedje** [comment]
When trying to compile this, I got `warning: layout did not converge within 5 attempts`. This explains why the counters are messed up (page numbers use the introspection features of Typst and if the convergence warning appears, there are no guarantees anymore).

The root cause seems to be
```typ
counter(page).display(i => if calc.odd(i) {
  pagebreak()
})
```
Today, this can be written as `pagebreak(to: "even")` which works properly.
